src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerMetricsDiagnosticListener.cs

```diff
@@ -29,41 +29,108 @@ internal sealed class HttpHandlerMetricsDiagnosticListener : ListenerHandler
 
         private readonly PropertyFetcher<HttpResponseMessage> stopResponseFetcher = new("Response");
         private readonly Histogram<double> httpClientDuration;
+        private readonly HttpClientInstrumentationMeterOptions options;
 
-        public HttpHandlerMetricsDiagnosticListener(string name, Meter meter)
+        public HttpHandlerMetricsDiagnosticListener(string name, Meter meter, HttpClientInstrumentationMeterOptions options)
             : base(name)
         {
+            this.options = options;
             this.httpClientDuration = meter.CreateHistogram<double>("http.client.duration", "ms", "measures the duration of the outbound HTTP request");
         }
 
         public override void OnEventWritten(string name, object payload)
         {
             if (name == OnStopEvent)
             {
-                if (Sdk.SuppressInstrumentation)
-                {
-                    return;
-                }
+                this.OnStopActivity(Activity.Current, payload);
+            }
+        }
 
-                var activity = Activity.Current;
-                if (this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) && response != null)
-                {
-                    var request = response.RequestMessage;
+        private void OnStopActivity(Activity activity, object payload)
+        {
+            if (Sdk.SuppressInstrumentation)
+            {
+                return;
+            }
+
+            if (!this.stopResponseFetcher.TryFetch(payload, out HttpResponseMessage response) || response == null)
+            {
+                // TODO: logging?
+                return;
+            }
+
+            var request = response.RequestMessage;
+            if (!this.TryFilterHttpRequestMessage(activity.OperationName, request))
+            {
+                return;
+            }
+
+            var tags = new List<KeyValuePair<string, object>>
```

Comment:
we should try to minimize heap allocations in hot path.

Taking a step back, we do not have any way to achieve "Enrich" equivalent in the SDK itself today, so this needs to be first design reviewed before making it part of the instrumentation library.

Reply:
Spoke to @vishweshbankwar offline about recent changes. We can do this for now (as this was already done for other instrumentation). Please make sure the allocation is not done by following similar approach as other instrumentation (aspnetcore).

Reply:
Hello, nice idea, I missed opportunity to create custom delegate while doing